Text doesn't load properly in Lockdown mode

vitorroriz opened the pull request:
#### a0c6667a0b87b8f7547a40186f41d003eb9f64fe
<pre>
Text doesn&apos;t load properly in Lockdown mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=284590">https://bugs.webkit.org/show_bug.cgi?id=284590</a>
<a href="https://rdar.apple.com/139031761">rdar://139031761</a>

Reviewed by Brent Fulgham.

We are doing two related things on this patch:

1. We were failing to recgonize that Safe Font Parser could not handle a font.

When `softLinkCoreTextCTFontManagerCreateMemorySafeFontDescriptorFromData` fails
to produce a font description we should return a null FontCustomPlatformData instead
of trying to create a FontCustomPlatformData object with it.

2. Avoiding fontLoaded re-entry:
When [1] happens we will fail to load the font and will deleted the related
extracted data. We do that by calling setErrorAndDeleteData, that calls CachedResource::error.
`error` notifies this resource client&apos;s so that fontLoaded is invoked for each of them. However,
this whole operation already started from a fontLoaded notification. So we are adding a flag
to avoid processing it again for the same client, in this case, CachedFontLoadRequest object.

* Source/WebCore/loader/cache/CachedFontLoadRequest.h:
* Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
(WebCore::FontCustomPlatformData::createMemorySafe):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Ahem-CFF.otf: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LockdownModeFonts.mm:
(TestWebKitAPI::TEST(LockdownMode, NotSupportedFontLoadingAPI)):

Canonical link: <a href="https://commits.webkit.org/287854@main">https://commits.webkit.org/287854@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/09327a18d8831d47152dbac48ebc569703f2e5dd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81108 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/633 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35051 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85637 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32094 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83218 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/651 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8440 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63319 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21082 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84177 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/410 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73824 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43617 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/307 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27986 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30552 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71836 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28544 "Found 1 new test failure: fast/editing/recursive-reapply-edit-command-crash.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87072 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8338 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5908 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71621 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8515 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69659 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70856 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17640 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14905 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13832 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8299 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13822 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8136 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11656 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9944 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/loader/cache/CachedFontLoadRequest.h:
```diff
@@ -94,6 +94,10 @@ class CachedFontLoadRequest final : public FontLoadRequest, public CachedFontCli
+        if (m_fontLoadedProcessed)
+            return;
+
+        m_fontLoadedProcessed = true;
```
Should we wait to initialize this until `m_fontLoadRequestClient->fontLoaded(*this)` returns?

brentfulgham reviewed (approved):
Thank you for getting this addressed so quickly!

vitorroriz committed:
Text doesn't load properly in Lockdown mode
https://bugs.webkit.org/show_bug.cgi?id=284590
rdar://139031761

Reviewed by Brent Fulgham.

We are doing two related things on this patch:

1. We were failing to recgonize that Safe Font Parser could not handle a font.

When `softLinkCoreTextCTFontManagerCreateMemorySafeFontDescriptorFromData` fails
to produce a font description we should return a null FontCustomPlatformData instead
of trying to create a FontCustomPlatformData object with it.

2. Avoiding fontLoaded re-entry:
When [1] happens we will fail to load the font and will deleted the related
extracted data. We do that by calling setErrorAndDeleteData, that calls CachedResource::error.
`error` notifies this resource client's so that fontLoaded is invoked for each of them. However,
this whole operation already started from a fontLoaded notification. So we are adding a flag
to avoid processing it again for the same client, in this case, CachedFontLoadRequest object.

* Source/WebCore/loader/cache/CachedFontLoadRequest.h:
* Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
(WebCore::FontCustomPlatformData::createMemorySafe):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Ahem-CFF.otf: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LockdownModeFonts.mm:
(TestWebKitAPI::TEST(LockdownMode, NotSupportedFontLoadingAPI)):

Canonical link: https://commits.webkit.org/287854@main
